### PR TITLE
start integration of madnlp

### PR DESCRIPTION
**Pull request by PierreMartinon**
docs bug when building locally ?

### CODE REVIEW

**Comment by PierreMartinon** [comment]
Benchmark at least works, even if madnlp seems slower.
Note that options are different than for ipopt, so we may still have to find the best ones for us.
```
julia> include("benchmark/benchmark.jl")
NLP Solver: madnlp   Linear solver: mumps
Blas config: LBTConfig([ILP64] libmkl_rt.so, [LP64] libmkl_rt.so)
Settings: tol=1e-08 grid_size=1000 precompile=true

Loading problems: beam bioreactor_1day_periodic fuller goddard jackson vanderpol 
Precompilation: beam bioreactor_1day_periodic fuller goddard jackson vanderpol 
Precompilation total time  38.01
Benchmark step
beam                           completed in   1.25 s after   28 iterations
bioreactor_1day_periodic       completed in  20.19 s after  373 iterations
fuller                         completed in   0.47 s after   21 iterations
goddard                        completed in   3.83 s after   31 iterations
jackson                        completed in   0.82 s after   23 iterations
vanderpol                      completed in   0.44 s after    3 iterations

Total time (s):  27.00
```

**Comment by PierreMartinon** [comment]
Code is still ugly, I'll keep working on it for a bit.
We do have 2 extensions now, for ipopt and madnlp.
However, we lost the custom error messages when trying to solve before loading the correct package: since we use a solver-specific struct now, if the package is missing the constructor will not be defined and will issue an error.

Function looks like this
```
# +++ some arguments are solver specific and should be defined only in the different calls to solve_docp ?
function solve(ocp::OptimalControlModel, description::Symbol...;
    init=CTDirect.__ocp_init_direct(),
    grid_size::Integer=CTDirect.__grid_size_direct(),
    display::Bool=CTDirect.__display(),
    print_level::Integer=CTDirect.__print_level_ipopt(),
    mu_strategy::String=CTDirect.__mu_strategy_ipopt(),
    max_iter::Integer=CTDirect.__max_iter(),
    tol::Real=CTDirect.__tol(),
    linear_solver::String=CTDirect.__linear_solver(),
    time_grid=nothing,
    kwargs...)

    method = getFullDescription(description, available_methods())
    #println(method)

    # build discretized OCP, including initial guess
    docp, nlp = direct_transcription(ocp, description, init=init, grid_size=grid_size, time_grid=time_grid)

    # override print_level if needed
    print_level = display ?  print_level : 0

    # solve DOCP
    # +++ using solver-specific structs will fail if package is not loaded...
    # +++ here set some specific default arguments, instead of above in solve() ?
    if :ipopt ∈ method
        solver = IpoptSolver(nlp)
        docp_solution = CTDirect.solve_docp(solver, docp, nlp, 
        print_level=print_level, mu_strategy=mu_strategy, tol=tol, max_iter=max_iter, linear_solver=linear_solver; kwargs...)
    elseif :madnlp ∈ method
        if print_level == 0
            print_level = MadNLP.ERROR
        else
            print_level = MadNLP.INFO
        end
        solver = MadNLPSolver(nlp, print_level=print_level)
        docp_solution = CTDirect.solve_docp(solver, docp, 
        tol=tol, max_iter=max_iter, linear_solver=linear_solver; kwargs...)
    else
        error("no known solver in method", method)
    end
 
    # build and return OCP solution
    return build_solution(docp, docp_solution)

 end
```

**Comment by jbcaillau** [comment]
> Code is still ugly, I'll keep working on it for a bit. We do have 2 extensions now, for ipopt and madnlp.
> 
> `build_solution` was redefined for madnlp since output format is not identical. The method for ipopt is still in CTDirect (ie not an extension) since it apparently returns a solution with a rather generic format, so it could act as the default one.

@PierreMartinon nice. regarding extension or not, I would say extension:
- no « default » as no solver should be loaded by default to keep `using OptimalControl` as light/fast as possible
- then one extension per supported solver
- it is nice to have MadNLP in the loop (although @frapac is on holiday 🥲)

**Comment by ocots** [comment]
Question: should `build_solution` be called `OptimalControlSolution` since it constructs an object of this type?

**Comment by PierreMartinon** [comment]
> Question: should `build_solution` be called `OptimalControlSolution` since it constructs an object of this type?

Indeed ! I'll have a look into it.

**Comment by PierreMartinon** [comment]
Updated benchmark as a function with arguments

Default ipopt/mumps
```
julia> bench();
Profile: NLP Solver ipopt with linear solver mumps
Blas config: LBTConfig([ILP64] libmkl_rt.so, [LP64] libmkl_rt.so)
Settings: tol=1e-08 grid_size=1000 precompile=true

Precompilation: beam bioreactor_1day fuller goddard jackson vanderpol 
Precompilation total time   2.63

beam                           completed in   0.66 s after   34 iterations
bioreactor_1day                completed in   5.82 s after  131 iterations
fuller                         completed in   0.37 s after   19 iterations
goddard                        completed in   2.15 s after   29 iterations
jackson                        completed in   0.64 s after   21 iterations
vanderpol                      completed in   0.40 s after    3 iterations

Total time (s):  10.03
```

Madnlp with default umfpack
```
julia> bench(nlp_solver=:madnlp);
Profile: NLP Solver madnlp with linear solver UmfpackSolver
Blas config: LBTConfig([ILP64] libmkl_rt.so, [LP64] libmkl_rt.so)
Settings: tol=1e-08 grid_size=1000 precompile=true

Precompilation: beam bioreactor_1day fuller goddard jackson vanderpol 
Precompilation total time   2.08

beam                           completed in   0.56 s after   29 iterations
bioreactor_1day                completed in   4.95 s after  127 iterations
fuller                         completed in   0.37 s after   21 iterations
goddard                        completed in   2.21 s after   30 iterations
jackson                        completed in   0.62 s after   23 iterations
vanderpol                      completed in   0.36 s after    3 iterations

Total time (s):   9.06
```

Madnlp / mumps (need using MadNLPMumps)
```
julia> bench(nlp_solver=:madnlp, linear_solver="MumpsSolver");
Profile: NLP Solver madnlp with linear solver MumpsSolver
Blas config: LBTConfig([ILP64] libmkl_rt.so, [LP64] libmkl_rt.so)
Settings: tol=1e-08 grid_size=1000 precompile=true

Precompilation: beam bioreactor_1day fuller goddard jackson vanderpol 
Precompilation total time   1.97

beam                           completed in   0.52 s after   29 iterations
bioreactor_1day                completed in   5.50 s after  114 iterations
fuller                         completed in   0.37 s after   21 iterations
goddard                        completed in   2.07 s after   30 iterations
jackson                        completed in   0.62 s after   23 iterations
vanderpol                      completed in   0.43 s after    3 iterations

Total time (s):   9.52
```

File: src/solve.jl [hunk 1]
```diff
@@ -67,13 +56,24 @@ Set initial guess in the DOCP
...
-    error("Please execute `using NLPModelsIpopt` before calling the solve method.")
+# placeholders (see CTSolveExt*** extensions)
+abstract type SolverTag end
+struct IpoptTag <: SolverTag end
+struct MadNLPTag <: SolverTag end
+
+function solve_docp(solver_tag, args...; kwargs...)
+    if typeof(solver_tag) == IpoptTag
+        #error("Please execute `using NLPModelsIpopt` before calling the solve method.")
+        throw(ExtensionError("Please execute `using NLPModelsIpopt`"))
```
Comment (ocots) [inline src/solve.jl:72]: You just have to write 

```julia
throw(ExtensionError(:NLPModelsIpopt))
```

File: test/suite/test_grid.jl [hunk 2]
```diff
@@ -66,14 +66,38 @@ end
...
     @test sol.objective ≈ 2.43 rtol=1e-2
 end
+
+# max t0 (free t0 and tf)
+ocp = Model(variable=true)
+state!(ocp, 2)
+control!(ocp, 1)
+variable!(ocp, 2)
+time!(ocp, ind0=1, indf=2)
+constraint!(ocp, :initial, lb=[0,0], ub=[0,0])
```
Comment (ocots) [inline test/suite/test_grid.jl:87]: If you want I have added `val` when `lb=ub`.

```julia
constraint!(ocp, :initial, val=[0, 0])
```

**Pull request merged**